### PR TITLE
Language switcher

### PR DESCRIPTION
**Pull request by zarino**
Language switcher for #676.

Closed:
closed

Open:
open

Feels a bit fiddly. Will revisit tomorrow morning with fresh eyes.

### CODE REVIEW

File: candidates/migrations/0022_create_standard_simple_fields.py [hunk 1]
```diff
@@ -39,7 +39,7 @@ def create_simple_fields(apps, schema_editor):
         ),
         SimpleField(
             name='gender',
-            label=u"Gender (e.g. “male”, “female”)",
+            label=u"Gender (e.g. male, female)",
```
Comment (zarino) [inline candidates/migrations/0022_create_standard_simple_fields.py]: Oops! This change shouldn’t have been included in this PR. Can someone remove it?

Comment (struan) [inline candidates/migrations/0022_create_standard_simple_fields.py]: I'll create a fixup when we merge all this in to remove this.

File: mysite/templates/generic_base.html [hunk 2]
```diff
@@ -73,12 +73,20 @@
...
-          </ul>
+          <form action="{% url 'set_language' %}" method="post">
+          {% csrf_token %}
+          <select name="language">
+            {% get_current_language as LANGUAGE_CODE %}
+            {% get_available_languages as LANGUAGES %}
+            {% get_language_info_list for LANGUAGES as languages %}
+            {% for language in languages %}
+            <option value="{{ language.code }}"{% if language.code == LANGUAGE_CODE %} selected="selected"{% endif %}>
+                {{ language.name_local }} ({{ language.code }})
```
Comment (wfdd) [inline mysite/templates/generic_base.html]: For semantic correctness, `language.name_local` should be marked up for `lang`, that is

```html
<span lang="{{ language.code }}"> {{ language.name_local }}</span> ({{ language.code }})
```

Screen readers may also be able to use this information to pronounce the name correctly.

**Comment by wfdd** [comment]
A couple of observations:

* `get_language_info` corrupts the language codes and names of languages not built into Django.  It works fine with Argentinian Spanish:
  ```py
In [1]: from django.utils import translation

In [2]: from django.conf import settings

In [3]: settings.LANGUAGES
Out[3]: 
[('el', 'Greek'),
 ('en', 'English'),
 ('es-ar', 'Argentinian Spanish'),
 ('fr', 'French'),
 ('cy-gb', 'Welsh'),
 ('es-cr', 'Costa Rican Spanish')]

In [4]: translation.get_language_info('es-ar')
Out[4]: 
{'bidi': False,
 'code': 'es-ar',
 'name': 'Argentinian Spanish',
 'name_local': 'español de Argentina'}
```
  But not the Costa Rican variety:
  ```py
In [5]: translation.get_language_info('es-cr')
Out[5]: {'bidi': False, 'code': 'es', 'name': 'Spanish', 'name_local': 'español'}
```
  I think it'd be best to bypass `get_language_info_list` and list our languages for translation in `mysite.extra_translations`.
* It should be possible to POST on click/select with XHR, serving the original dropdown to clients that support JavaScript.  Having to press 'Go' is not ideal.
* Perhaps in a separate PR, it should be made possible to override the `User-Agent` locale as the default locale.

**Comment by zarino** [comment]
> It should be possible to POST on click/select with XHR, serving the original dropdown to clients that support JavaScript. Having to press 'Go' is not ideal.

Already working on it! 👍

File: mysite/templates/generic_base.html (hunk 2 above)
Comment (zarino) [inline mysite/templates/generic_base.html]: Good spot!

**Comment by zarino** [comment]
@struan @mhl: Commit 31bb5384af1e215ff7acb74ba07cf19924dc2840 reverts all the custom dropdown menu styling and instead styles up the `<select>` element. If JavaScript is available, the submit button is hidden, and the form is automatically submitted whenever you change the `<select>` value.

language-switcher

**Comment by mhl** [comment]
I think [@wfdd's first point](https://github.com/mysociety/yournextrepresentative/pull/917#issuecomment-219490964) still needs to be addressed here - I'm assuming that's why the following fails:

* (In chrome) set your first language to be Greek
* Reload the front page, text appears in Greek as expected
* Select Welsh from the language drop-down
* The form is POSTed, the page reloads as expected, but the text is still in Greek <-- the bug
* Select Argentinian Spanish from the language drop-down
* The form is POSTed, the page reloads as expected, and the text is in Argentinian Spanish

File: mysite/settings/conf.py [hunk 3]
```diff
@@ -59,6 +60,27 @@ def get_settings(conf_file_leafname, election_app=None, tests=False):
...
+        'es-cr': {
+            'bidi': False,
+            'code': 'es-cr',
+            'name': 'Spanish',
+            'name_local': u'español de Costa Rica',
+        },
+    }
+
+    LANG_INFO = dict(locale.LANG_INFO.items() + EXTRA_LANG_INFO.items())
+    locale.LANG_INFO = LANG_INFO
```
Comment (mhl) [inline mysite/settings/conf.py]: It'd be neater to do `locale.LANG_INFO.update(EXTRA_LANG_INFO)` - I think the effect would be the same.

Comment (wfdd) [inline mysite/settings/conf.py]: You can't concat dict.items in Python 3 anyway.

**Pull request merged**